### ISSUE TITLE
mysqli example isn't working

### ISSUE THREAD

**Issue by c9s** [krakjoe/pthreads#674]
Environment
==========

  * PHP: 7.1.0
  * pthreads: master branch
  * OS: OS X

Summary
======

Running the mysqli worker example returns exit code 0, however the result was not dumped.

It seems that `collect` method was not called:

```
$pool->collect(function($query) {
    die('test');
    printf($done = $query->getResult());
    return count($done);
});
```

Reproducing Code
==============

```
$pool->collect(function($query) {
    die('test');
    printf($done = $query->getResult());
    return count($done);
});
```

Expected Output
=============

    test

Actual Output
============

empty result.

**Comment by c9s** [comment]
Here is the working version:

```php
/**
* This example illustrates best practice with regard to using MySQLi in multiple threads
*
* For convenience and simplicity it uses a Pool.
**/
class Connect extends Worker {
    public function __construct($hostname, $username, $password, $database, $port = 3306)
    {
        $this->hostname = $hostname;
        $this->username = $username;
        $this->password = $password;
        $this->database = $database;
        $this->port     = $port;
    }
    
    public function getConnection() {
        if (!self::$link) {
            self::$link = new mysqli(
                $this->hostname, 
                $this->username, 
                $this->password, 
                $this->database, 
                $this->port);
        }
        
        /* do some exception/error stuff here maybe */       
        return self::$link;
    }
    
    protected $hostname;
    protected $username;
    protected $password;
    protected $database;
    protected $port;
    
    /**
    * Note that the link is stored statically, which for pthreads, means thread local
    **/
    protected static $link;
}

class QueryJob extends Threaded {

    protected $done;

    public function __construct($sql) {
        $this->sql = $sql;
    }
    
    public function run() {
        $mysqli = $this->worker->getConnection();
        
        $result = $mysqli->query($this->sql);
        
        if ($result) {    
            while (($row = $result->fetch_assoc())) {
                $rows[] = $row;
            }
        }
        $this->result = $rows;
        $this->done = true;
    }

    public function getDone() {
        return $this->done;
    }
    
    public function getResult() {
        return $this->result;
    }
    
    protected $sql;
    protected $result;
}

$pool = new Pool(4, "Connect", ["localhost", "root", "", "mysql"]);
$pool->submit(new QueryJob("SHOW PROCESSLIST;"));
$pool->submit(new QueryJob("SHOW PROCESSLIST;"));
$pool->submit(new QueryJob("SHOW PROCESSLIST;"));
$pool->submit(new QueryJob("SHOW PROCESSLIST;"));
$pool->submit(new QueryJob("SHOW PROCESSLIST;"));
$pool->submit(new QueryJob("SHOW PROCESSLIST;"));

/* ::collect is used here for shorthand to dump query results */

while($pool->collect(function ($job) {
    return $job->getDone();
})) continue;

var_dump($pool);

$pool->shutdown();
```

**Comment by c9s** [comment]
It seems that the collector can't fetch any non-scalar value from the job. but boolean works.

**Comment by c9s** [comment]
The returned row array seems to be corrupted

```
array(8) {
  ["Id"]=>
  string(3) "234"
  ["User"]=>
    string(4) "rootstring(5) ""
mysql  [""
Host  [""]=>
Command  "]=>
string(9) "  localhoststring(5) ""
Query  [""
db  [""]=>
Time  "]=>
string(5) "  mysql"
  ["Command"]=>
  string(5) "Query"
  ["Time"]=>
  string(1) "0"
  ["State"]=>
  string(4) "init"
  ["Info"]=>
  string(16) "SHOW PROCESSLIST"
}
```

**Comment by c9s** [comment]
I updated the example to use Worker::collect, however the retrieved rows can't be fetched from the main process.

I traced a little bit and found when it creates pthread store it uses `pthreads_store_tostring` for array,  however the retrieved rows (arrays) are stored in an array (nested array)

And when the value is `IS_ARRAY`, `zend_array_dup` is called to duplicate the array, in which, it only copies the first level element.

My suspicion is, for the nested arrays, it only saves the references, and so the array is corrupted when the thread finished the work.

**Comment by c9s** [comment]
Here is the code that could reproduce the issue:

```php
<?php
/**
* This example illustrates best practice with regard to using MySQLi in multiple threads
*
* For convenience and simplicity it uses a Pool.
**/
class MysqliWorker extends Worker {

    public function __construct($hostname, $username, $password, $database, $port = 3306)
    {
        $this->hostname = $hostname;
        $this->username = $username;
        $this->password = $password;
        $this->database = $database;
        $this->port     = $port;
    }
    
    public function getConnection() {
        if (!self::$link) {
            self::$link = new mysqli(
                $this->hostname, 
                $this->username, 
                $this->password, 
                $this->database, 
                $this->port);
        }
        
        /* do some exception/error stuff here maybe */       
        return self::$link;
    }
    
    protected $hostname;
    protected $username;
    protected $password;
    protected $database;
    protected $port;
    
    /**
    * Note that the link is stored statically, which for pthreads, means thread local
    **/
    protected static $link;
}

class QueryJob extends Threaded {

    public function __construct($sql) {
        $this->sql = $sql;
    }
    
    public function run()
    {
        $mysqli = $this->worker->getConnection();
        $result = $mysqli->query($this->sql);
        $rows = [];
        if ($result) {    
            while (($row = $result->fetch_assoc())) {
                $rows[] = $row;
            }
        }
        // $this->result = serialize($rows);
        $this->result = $rows;
        $this->g = true;
    }

    public function getResult()
    {
        return $this->result;
        // return unserialize($this->result);
    }


    public function setGarbage()
    {
        return $this->g = true;
    }

    public function isGarbage() : bool {
        return $this->g;
    }
    
    public $g = false;
    public $sql;
    public $result;
}

$jobs = [];
$worker = new MysqliWorker('localhost', 'root', null, 'mysql', 3306);
$worker->start();
$worker->stack($jobs[] = new QueryJob("SHOW PROCESSLIST;"));
$worker->stack($jobs[] = new QueryJob("SHOW PROCESSLIST;"));
$worker->stack($jobs[] = new QueryJob("SHOW PROCESSLIST;"));
$worker->stack($jobs[] = new QueryJob("SHOW PROCESSLIST;"));


while($worker->collect()) continue;

foreach ($jobs as $job) {
    var_dump($job->getResult());
}

/*
while($worker->collect(function($job){
    var_dump($job->getResult()); 
    return $job->setGarbage();
})) continue;
*/

var_dump($worker->getStacked());


$worker->shutdown();

```

**Comment by krakjoe** [comment]
It's just a bad example, from pthreads v2 ... you are not meant to use `collect` for that purpose, it has non-deterministic behaviour and so is not suitable for that.

I don't have time to write a better example right now, so I'll leave this open until I do ... but there's no bug here.

**Comment by c9s** [comment]
@krakjoe thanks for the reply!

I replaced the collect method with join call yesterday, and still, the same message. Please see my #675 Pull request. 

Thank you for all the work!

**Comment by tpunt** [comment]
This has now been fixed in 6832534.
